Sources/System/MachPort.swift

```diff
@@ -0,0 +1,327 @@
+/*
+ This source file is part of the Swift System open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+#if $MoveOnly && (os(macOS) || os(iOS) || os(watchOS) || os(tvOS))
+
+import Darwin.Mach
+
+public protocol MachPortRight {}
+
+private func machPrecondition(
+    file: StaticString = #file,
+    line: UInt = #line,
+    _ body: @autoclosure () -> kern_return_t
+) {
+    let kr = body()
+    let expected = KERN_SUCCESS
+    precondition(kr == expected, file: file, line: line)
+}
+
+public enum Mach {
+    @_moveOnly
+    public struct Port<RightType: MachPortRight> {
+        internal var name: mach_port_name_t
+        internal var context: mach_port_context_t
+
+        /// Transfer ownership of an existing unmanaged Mach port right into a
+        /// Mach.Port by name.
+        ///
+        /// This initializer aborts if name is MACH_PORT_NULL, or if name is
+        /// MACH_PORT_DEAD and the type T of Mach.Port<T> is Mach.ReceiveRight.
+        ///
+        /// If the type of the right does not match the type T of Mach.Port<T>
+        /// being constructed, behavior is undefined.
```

Comment:
I agree to keep the term "undefined" aligned with C's notion of "undefined behavior", e.g. memory unsafety. What happens here, is it merely an unspecified result? For example, undefined might read from memory out of bounds while an unspecified result might throw errors when you try to use it without violating memory safety.

Reply:
My initial reading triggered all the alarms for me because I thought it actually meant undefined behavior in the C sense, and we really don't want that. Now I simply don't know what the sentence means. It would be advantageous to clarify this as fully as possible!

Reply:
The current interface is simple enough that you are most likely to hit unexpected traps for things that don't throw and are expected to always succeed for the correct type (in general every `machPrecondition`), but things may also silently succeed if your task does hold a right of that type at the time of the operation. However, some operations may have more complex failure modes, especially when the interface grows to encompass more complex things like messaging. e.g. If you created a `Mach.Port<Mach.SendRight>` with a send-once right, merely sending a message over that right consumes the right, which would cause downstream operations to trap.